app/org/sagebionetworks/bridge/services/OAuthService.java

```diff
@@ -0,0 +1,125 @@
+            int pageSize, String offsetKey) {
+        checkNotNull(study);
+        checkNotNull(vendorId);
+        
+        // Verify the provider exists
+        OAuthProvider provider = study.getOAuthProviders().get(vendorId);
+        if (provider == null) {
+            throw new EntityNotFoundException(OAuthProvider.class);
+        }
+        ForwardCursorPagedResourceList<OAuthAccessGrant> list = grantDao.getAccessGrants(study, vendorId,
+                offsetKey, pageSize);
+
+        List<String> healthCodes = list.getItems().stream().map(OAuthAccessGrant::getHealthCode)
+                .collect(Collectors.toList());
+        
+        return new ForwardCursorPagedResourceList<String>(healthCodes, list.getNextPageOffsetKey())
+                .withRequestParam(PAGE_SIZE, pageSize)
+                .withRequestParam(OFFSET_KEY, offsetKey);
+    }
+    
+    public OAuthAccessToken requestAccessToken(Study study, String healthCode, OAuthAuthorizationToken authToken) {
+        checkNotNull(study);
+        checkNotNull(healthCode);
+        checkNotNull(authToken);
+        
+        return retrieveAccessToken(study, authToken.getVendorId(), healthCode, authToken);
+    }
+    
+    public OAuthAccessToken getAccessToken(Study study, String vendorId, String healthCode) {
+        checkNotNull(study);
+        checkNotNull(vendorId);
+        checkNotNull(healthCode);
+        
+        return retrieveAccessToken(study, vendorId, healthCode, null);
+    }
+    
+    private OAuthAccessToken retrieveAccessToken(Study study, String vendorId, String healthCode,
+            OAuthAuthorizationToken authToken) {
+        checkNotNull(study);
+        checkNotNull(vendorId);
+        checkNotNull(healthCode);
+        
+        OAuthProvider provider = study.getOAuthProviders().get(vendorId);
+        if (provider == null) {
+            throw new EntityNotFoundException(OAuthProvider.class);
+        }
+        OAuthAccessGrant grant = null;
+        
+        try {
+            // If client has submitted an authorization token, we always refresh the grant
+            if (authToken != null && authToken.getAuthToken() != null) {
+                grant = providerService.requestAccessGrant(provider, authToken);
+            } else {
+                // If not, start first by seeing if a grant has been saved
+                grant = grantDao.getAccessGrant(study.getStudyIdentifier(), vendorId, healthCode);
+            }
+            // If no grant was saved or successfully returned from a grant, it's not found.
+            if (grant == null) {
+                throw new EntityNotFoundException(OAuthAccessGrant.class);
+            } else if (getDateTime().isAfter(grant.getExpiresOn())) {
+                // If there's a grant record, but it has expired, attempt to refresh it
+                grant = providerService.refreshAccessGrant(provider, vendorId, grant.getRefreshToken());
```

Comment:
If this part fails, a mechanism to get a new grant from the participant would be useful